### PR TITLE
A command to show when users in a given room have joined

### PR DESCRIPTION
**Pull request by Yoric**
This should be useful during spam attacks.

### CODE REVIEW

File: src/utils.ts [hunk 1]
```diff
@@ -416,3 +416,228 @@ export function patchMatrixClient() {
     patchMatrixClientForConciseExceptions();
     patchMatrixClientForRetry();
 }
+
+/**
+ * Common duration units, used for commands that accept durations such as "1 day".
+ */
+export enum DurationUnit {
```
Comment (Gnuxie) [inline src/utils.ts:423]: I feel uncomfortable with putting in `src/utils.ts` this massive 200 lines long error prone duration handling code given we've had problems with rolling our own time stuff in other projects

Comment (Yoric) [reply src/utils.ts:423]: What would you suggest?

Comment (Yoric) [reply src/utils.ts:423]: We could also use https://www.npmjs.com/package/parse-duration, I suppose.

Comment (qbit) [reply src/utils.ts:423]: fwiw, https://github.com/wanasit/chrono seems to work fairly well in logseq (speaking as an end user of logseq)

Comment (Yoric) [reply src/utils.ts]: Ah, well, I've used parse-duration + humanize-duration.

File: src/RoomMembers.ts [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+        this._joinsByTimestamp = this._joinsByTimestamp.filter(join => this.isStillValid(join));
+        this._leaves = new Map();
+    }
+
+    /**
+     * Determine whether a `join` is still valid or has been superseded by a `leave`.
+     *
+     * @returns true if the `join` is still valid.
+     */
+    private isStillValid(join: Join): boolean {
```
Comment (jesopo) [inline src/RoomMembers.ts:90]: curiousity; how do we end up processing a join after a leave?

File: src/RoomMembers.ts [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+     *
+     * @returns a `Date` if the user is currently in the room and has joined
+     * since the start of Mjölnir, `null` otherwise.
+     */
+    public get(userId: string): Date | null {
+        let ts = this._joinsByUser.get(userId);
+        if (!ts) {
+            return null;
+        }
+        return new Date();
```
Comment (jesopo) [inline src/RoomMembers.ts:150]: shouldn't this be the `Date` that the user joined?

File: src/RoomMembers.ts [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+            return;
+        }
+        const userId = event['state_key'];
+        if (!userId) {
+            // Ill-formed event.
+            return;
+        }
+
+        const userState = event['content']['membership'];
+        let prevMembership = event['unsigned']?.['prev_content']?.['membership'] || "leave";
```
Comment (jesopo) [inline src/RoomMembers.ts:247]: shortcoming of my matrix knowledge; why is previous state important?

**Comment by rubo77** [comment]
Why is this useful?

File: src/RoomMembers.ts (hunk 3 above)
Comment (Yoric) [reply src/RoomMembers.ts:150]: Oops! Good catch!

File: src/RoomMembers.ts (hunk 4 above)
Comment (Yoric) [reply src/RoomMembers.ts:247]: A user changing display name will re-trigger a `join`, if my memory serves. A user may also be banned after leaving, but this doesn't change the fact that it has already left. There may be other non-trivial transitions.

See https://spec.matrix.org/v1.2/client-server-api/#mroommember .

**Comment by Yoric** [comment]
> Why is this useful?

We've had cases of rooms being actively attacked. It's useful for moderators to know who has joined recently, for instance to perform a mass kick of every account suspected to be a bot (and then reinvite users one by one).

File: src/RoomMembers.ts (hunk 2 above)
Comment (jesopo) [reply src/RoomMembers.ts:90]: what would make that "invalid"?

Comment (Yoric) [reply src/RoomMembers.ts:90]: In that case, we're checking whether there has been a `leave` after a `join`. If so, the `join` date isn't valid anymore.

**Review by jesopo — APPROVED** [review]

**Pull request merged**